File: analytics/metrics-platform/src/main/java/org/wikimedia/metrics_platform/config/curation/CurationRules.java
package org.wikimedia.metrics_platform.config.curation;

import java.util.Collection;
import java.util.function.Predicate;

import javax.annotation.ParametersAreNullableByDefault;

import com.google.gson.annotations.SerializedName;

import lombok.Builder;
import lombok.Value;

@Builder @Value
@ParametersAreNullableByDefault
public class CurationRules<T> implements Predicate<T> {
    @SerializedName("equals") T isEquals;
    @SerializedName("not_equals") T isNotEquals;
    Collection<T> in;
    @SerializedName("not_in") Collection<T> notIn;

    @Override
    public boolean test(T value) {
        if (value == null) return false;

        return (isEquals == null || isEquals.equals(value))
                && (isNotEquals == null || !isNotEquals.equals(value))
                && (in == null || in.contains(value))
                && (notIn == null || !notIn.contains(value));
    }
}

File: analytics/metrics-platform/src/main/java/org/wikimedia/metrics_platform/config/curation/CurationRules.kt
package org.wikimedia.metrics_platform.config.curation

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.function.Predicate

@Serializable
class CurationRules<T> : Predicate<T?> {
    @SerialName("equals") var isEquals: T? = null

    @SerialName("not_equals") var isNotEquals: T? = null

    @SerialName("in") var isIn: MutableCollection<T?>? = null

    @SerialName("not_in") var isNotIn: MutableCollection<T?>? = null

    override fun test(value: T?): Boolean {
        if (value == null) return false
        return (isEquals == null || isEquals == value)
                && (isNotEquals == null || isNotEquals != value)
                && (isIn == null || isIn!!.contains(value))
                && (isNotIn == null || !isNotIn!!.contains(value))
    }
}
